src/coreclr/jit/block.cpp

```diff
@@ -1278,7 +1278,7 @@ BasicBlock* BasicBlock::GetSucc(unsigned i, Compiler* comp)
             return bbJumpDest;
 
         case BBJ_NONE:
-            return bbNext;
+            return GetFallThroughSucc();
```

Comment:
Here and in a number of other places in this change, the pattern seems to be to change: `BBJ_NONE -> bbNext / BBJ_COND -> bbNext+bbJumpDest` to `BBJ_NONE -> GetFallThroughSucc() / BBJ_COND -> GetFallThroughSucc()+bbJumpDest`. The `BBJ_COND` part makes sense, but the `BBJ_NONE` changes confuse me.

If `GetFallThroughSucc()` is the normal successor of `BBJ_COND`, why touch `BBJ_NONE`?

Side note: I think you will also want to update `VisitAllSuccs/VisitRegularSuccs`.

Reply:
I'm not sure what @AndyAyersMS's plan is here, but a "lazy" approach to reordering the block layout might be moving a block without considering it might be a `BBJ_NONE` and its successor is its old `bbNext`, and just converting the `BBJ_NONE` to `BBJ_ALWAYS` after the reorder. If we do something like this, I think it would be nice to always use `GetFallThroughSucc` for `BBJ_NONE` blocks so we can assert the fall-through successor matches `bbNext`; if we mess something up during block reordering, hopefully this will help us catch it.

> Side note: I think you will also want to update `VisitAllSuccs/VisitRegularSuccs`.
Thanks for catching that. I'll fix those.

Reply:
I feel there is some overloading of meanings for "the fall-through successor".

The change at large is needed because in the refactored flowgraph, the "false" successor of `BBJ_COND` (and possibly the `BBJ_ALWAYS` part of callfinally pairs) will be distinct from `bbNext`. So, overall, the imported blocks will change as follows:
1) `BBJ_NONE -> bbNext` -> `BBJ_ALWAYS -> bbJumpDest` (can be reordered arbitrarily later on).
2) `BBJ_COND -> bbNext, bbJumpDest` -> `BBJ_COND -> bbTheNewFieldBeingAdded, bbJumpDest` (same).

In other words, "fall through" will not exist as an explicit flowgraph concept like it does today (perhaps until after a very late layout stage, where `BBJ_NONE` may appear as an optional optimization).

In that light, I would expect changes to basically replace `case BBJ_COND: if (<false branch>) return bbNext;` with `case BBJ_COND: if (<false branch>) return bbTheNewFieldBeingAdded;`, and indeed they do in some places like switch recognition, but not in others.

Reply:
My initial plan was to disallow `BBJ_NONE` until we'd set block layout. I think we'll have to chip away at this as we have a lot of code that needs to work both before and after we do layout.

So a plausible starting point is to allow `BBJ_NONE` but treat it somewhat like `BBJ_ALWAYS`, meaning it has to explicitly name its flow successor, and find some way to start incrementally removing the early `BBJ_NONE` cases.

Reply:
I wonder if simply not producing `BBJ_NONE` (anywhere), instead producing `BBJ_ALWAYS`, and adding a small peephole to codegen to not emit "branch to next block" would be a faster/easier way to go about this.

Reply:
> I wonder if simply not producing BBJ_NONE (anywhere), instead producing BBJ_ALWAYS, and adding a small peephole to codegen to not emit "branch to next block" would be a faster/easier way to go about this.

With this approach, maybe we can get rid of `bbFallThroughSucc` and instead add a new member specific to `BBJ_COND` for its false branch target.